contrib/http_wasm/filters/http/source/vm.h

```diff
@@ -0,0 +1,123 @@
+  void registerCallbacks() override;
+  void getFunctions() override;
+
+  // AccessLog::Instance
+  void log(const PluginSharedPtr& plugin, const Http::RequestHeaderMap* request_headers,
+           const Http::ResponseHeaderMap* response_headers,
+           const Http::ResponseTrailerMap* response_trailers,
+           const StreamInfo::StreamInfo& stream_info, AccessLog::AccessLogType access_log_type);
+
+  void initializeLifecycle(Server::ServerLifecycleNotifier& lifecycle_notifier);
+
+protected:
+  friend class Context;
+
+  Stats::ScopeSharedPtr scope_;
+  Api::Api& api_;
+  Stats::StatNamePool stat_name_pool_;
+  const Stats::StatName custom_stat_namespace_;
+  Upstream::ClusterManager& cluster_manager_;
+  Event::Dispatcher& dispatcher_;
+  Event::PostCb server_shutdown_post_cb_;
+  absl::flat_hash_map<uint32_t, Event::TimerPtr> timer_; // per root_id.
+  TimeSource& time_source_;
+
+  // // Lifecycle stats
+  Common::Wasm::LifecycleStatsHandler lifecycle_stats_handler_;
+};
+using WasmSharedPtr = std::shared_ptr<Wasm>;
+
+class WasmHandle : public Host::WasmHandleBase, public ThreadLocal::ThreadLocalObject {
+public:
+  explicit WasmHandle(const WasmSharedPtr& wasm)
+      : Host::WasmHandleBase(std::static_pointer_cast<Host::WasmBase>(wasm)), wasm_(wasm) {}
+
+  WasmSharedPtr& wasm() { return wasm_; }
+
+private:
+  WasmSharedPtr wasm_;
+};
+using WasmHandleSharedPtr = std::shared_ptr<WasmHandle>;
+
+class PluginHandle : public Host::PluginHandleBase {
+public:
+  explicit PluginHandle(const WasmHandleSharedPtr& wasm_handle, const PluginSharedPtr& plugin)
+      : Host::PluginHandleBase(std::static_pointer_cast<Host::WasmHandleBase>(wasm_handle),
+                               std::static_pointer_cast<Host::PluginBase>(plugin)),
+        plugin_(plugin), wasm_handle_(wasm_handle) {}
+
+  WasmHandleSharedPtr& wasmHandle() { return wasm_handle_; }
+  uint32_t rootContextId() { return wasm_handle_->wasm()->getRootContext()->id(); }
+
+private:
+  PluginSharedPtr plugin_;
+  WasmHandleSharedPtr wasm_handle_;
+};
+
+using PluginHandleSharedPtr = std::shared_ptr<PluginHandle>;
+class PluginHandleSharedPtrThreadLocal : public ThreadLocal::ThreadLocalObject {
+public:
+  PluginHandleSharedPtrThreadLocal(PluginHandleSharedPtr handle) : handle_(handle){};
+  PluginHandleSharedPtr& handle() { return handle_; }
+
+private:
+  PluginHandleSharedPtr handle_;
+};
+
+using CreateWasmCallback = std::function<void(WasmHandleSharedPtr)>;
+// Returns false if createWasm failed synchronously. This is necessary because xDS *MUST* report
+// all failures synchronously as it has no facility to report configuration update failures
+// asynchronously. Callers should throw an exception if they are part of a synchronous xDS update
+// because that is the mechanism for reporting configuration errors.
+bool createWasm(const PluginSharedPtr& plugin, const Stats::ScopeSharedPtr& scope,
```

Comment:
so one of the crazy glitches in Proxy-Wasm impl was it has to support HTTP fetching, which introduced extremely complex code in this `createWasm` function, but it has never been used due to the shitty failure mode which is impossible to recover.  In fact, in Istio, the istio-agent fetches the Wasm binary instead, saves it locally, and passes the local pass to the Envoy, rather than relying on Envoy-level HTTP fetching. 

So I think you can simplify `createWasm` function (I strongly believe we should rename to `createVM` or something like that btw) where just reads the local path and create the VM synchronously. That way we don't need this callBack style-impossible-to-unittest complexity in here.

Reply:
done